analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R

```diff
@@ -0,0 +1,84 @@
+library(dplyr)
+library(Seurat)
+library(ggpubr)
+pre_seuratobj <- function(obj, nfeatures = 500, run_harmony = TRUE, reduction = "harmony", ndims = 50){
+  ######## Normalize, scale, feature selection
+  obj <- Seurat::NormalizeData(obj, normalization.method = "LogNormalize")
+  obj <- Seurat::FindVariableFeatures(obj, selection.method = "vst", nfeatures = nfeatures)
+  obj <- Seurat::ScaleData(obj, features = Seurat::VariableFeatures(object = obj))
+  # obj <- Seurat::SCTransform(obj)
+  obj <- Seurat::RunPCA(obj, features = Seurat::VariableFeatures(object = obj))
+  # Seurat::ElbowPlot(obj, ndims = 50)
+  
+  ######## batch effect correction
+  if (run_harmony){
+    obj <- harmony::RunHarmony(obj, group.by.vars = "library_id")
+  }
+  
+  ######## Clustering and dimentional reduction
+  obj <- Seurat::FindNeighbors(obj, dims = 1:ndims, reduction = reduction)
+  obj <- Seurat::FindClusters(obj, resolution = 0.8, algorithm = 1)
+  obj <- Seurat::RunUMAP(obj, dims = 1:ndims, reduction = reduction)
+  
+  return(obj)
+}
+
+## remove duplicated genes in the list for dot plot
+clean_gslist <- function(inlist, obj){
+  dupgenes <- unique(unlist(inlist)[duplicated(unlist(inlist))])
+  availgenes <- rownames(obj)
+  for (i in 1:length(inlist)) {
+    element <- inlist[[i]][inlist[[i]] %in% availgenes]
+    element <- element[!element %in% dupgenes]
+    inlist[[i]] <- element
+  }
+  return(inlist)
+}
+
+## create and process a seurat object from provided sce object
+process_sce <- function( sample, library,
+                         path_proj = "/home/lightsail-user/git/OpenScPCA-analysis/data/current/SCPCP000014",
+                         db_proj = "/home/lightsail-user/git/OpenScPCA-analysis/data/current/results/doublet-detection/SCPCP000014",
+                         path_anal = "/home/lightsail-user/git/OpenScPCA-analysis/analyses/cell-type-wilms-tumor-14/"
+){
+  
+  ########### per sample rds preprocessing ##########
+  
+  #sample <- "SCPCS000517"; library <- "SCPCL000849"
+  rds <- readRDS(paste0(path_proj,"/",sample,"/",library,"_processed.rds"))
+  # out_rds <- paste0(path_proj,"/",sample,"/",library,"_processed_genesym.rds")
+  db <- read.table(paste0(db_proj,"/",sample,"/",library,"_processed_scdblfinder.tsv"), header = T)
+  
+  ######## convert ensembl IDs to symbols and dedup
+  rownames(rds) <- rds@rowRanges@elementMetadata@listData[["gene_symbol"]]
+  rds <- singleCellTK::dedupRowNames(rds, as.rowData = F, return.list = F)
+  rds <- rds[!is.na(rownames(rds)),]
+  rds <- rds[!duplicated(rownames(rds)),]
+  # readr::write_rds(rds, out_rds, compress = "bz2")
```

Comment:
> I expect that part of the reason that you are taking this approach is that the marker genes that you will be using later are listed by gene symbol, and if that is the case, I might try converting in the other direction?

I converted to gene symbol because most of reference databases (e.g Human kidney atlas, other public single-cell datasets I used in my test) have gene symbols as row names by default. And for sure, I used gene symbol when curating markers genes. Thus, when testing different methods, it's always more convenient to use gene symbol.

On the other hand, I agree that simply converting to gene symbol and removing duplicates may lose some information, since it's less standard than Ensembl IDs.

To align with your project standard naming, I would keep ensembl IDs as default for this step. I'll look more into reference databases & marker gene list, trying to convert gene symbol to ensembl IDs. Or converting row names to gene symbols for individual sections which don't support ensembl IDs.

> I am a bit confused about why there are still duplicated rows after running singleCellTK::dedupRowNames. Do you have a thought there?

Never mind. Looking back to my scripts, I realize it's redundant. I think kept it just to avoid any error messages.